python/lsst/pipe/base/pipelineTask.py

```diff
@@ -389,35 +447,35 @@ def runQuantum(self, quantum, butler):
         # get all data from butler
         inputDataIds = {}
         inputs = {}
-        for key, value in self.config.items():
-            if isinstance(value, InputDatasetConfig):
-                dataRefs = quantum.predictedInputs[value.name]
-                dataIds = [dataRef.dataId for dataRef in dataRefs]
-                data = [butler.get(dataRef) for dataRef in dataRefs]
-                if value.scalar:
-                    # unpack single-item lists
-                    if len(dataRefs) != 1:
-                        raise ScalarError(key, len(dataRefs))
-                    data = data[0]
-                    dataIds = dataIds[0]
-                inputDataIds[key] = dataIds
-                inputs[key] = data
+        descriptors = self.getInputDatasetTypes(self.config)
+        for key, descriptor in descriptors.items():
+            dataRefs = quantum.predictedInputs[descriptor.datasetType.name]
+            dataIds = [dataRef.dataId for dataRef in dataRefs]
+            data = [butler.get(dataRef) for dataRef in dataRefs]
+            if descriptor.scalar:
+                # unpack single-item lists
+                if len(dataRefs) != 1:
+                    raise ScalarError(key, len(dataRefs))
+                data = data[0]
+                dataIds = dataIds[0]
+            inputDataIds[key] = dataIds
+            inputs[key] = data
 
         # lists of DataRefs/DataIds for output datasets
         outputDataRefs = {}
         outputDataIds = {}
-        for key, value in self.config.items():
-            if isinstance(value, OutputDatasetConfig):
-                dataRefs = quantum.outputs[value.name]
-                dataIds = [dataRef.dataId for dataRef in dataRefs]
-                if value.scalar:
-                    # unpack single-item lists
-                    if len(dataRefs) != 1:
-                        raise ScalarError(key, len(dataRefs))
-                    dataRefs = dataRefs[0]
-                    dataIds = dataIds[0]
-                outputDataRefs[key] = dataRefs
-                outputDataIds[key] = dataIds
+        descriptors = self.getOutputDatasetTypes(self.config)
+        for key, descriptor in descriptors.items():
+            dataRefs = quantum.outputs[descriptor.datasetType.name]
+            dataIds = [dataRef.dataId for dataRef in dataRefs]
+            if descriptor.scalar:
+                # unpack single-item lists
+                if len(dataRefs) != 1:
+                    raise ScalarError(key, len(dataRefs))
+                dataRefs = dataRefs[0]
+                dataIds = dataIds[0]
+            outputDataRefs[key] = dataRefs
+            outputDataIds[key] = dataIds
```

Comment:
This block for outputs and the one above for inputs are awfully similar.  Could you most of this into a separate method?  Could even be a local function.